server/game/gamesteps/conflict/conflictflow.js

```diff
@@ -97,29 +115,69 @@ class ConflictFlow extends BaseStepWithPipeline {
             return;
         }
 
-        _.each(this.conflict.attackers, card => card.inConflict = true);
-        this.game.recordConflict(this.conflict);
-    }
+        let provinceSlot = this.conflict.conflictProvince ? this.conflict.conflictProvince.location : Locations.ProvinceOne;
+        let provinceName = (this.conflict.conflictProvince && this.conflict.conflictProvince.facedown) ? provinceSlot : this.conflict.conflictProvince;
+        this.game.addMessage('{0} is initiating a {1} conflict at {2}, contesting {3}', this.conflict.attackingPlayer, this.conflict.conflictType, provinceName, this.conflict.ring);
 
-    payAttackerCosts() {
-        const totalFateCost = this.conflict.attackers.reduce((total, card) => total + card.sumEffects(EffectNames.FateCostToAttack), 0);
-        if(!this.conflict.conflictPassed && totalFateCost > 0) {
-            this.game.addMessage('{0} pays {1} fate to declare his attackers', this.conflict.attackingPlayer, totalFateCost);
-            const costEvents = [];
-            Costs.payFate(totalFateCost).addEventsToArray(costEvents, this.game.getFrameworkContext(this.conflict.attackingPlayer));
-            this.game.openEventWindow(costEvents);
-        }
+        const params = {
+            conflict: this.conflict,
+            type: this.conflict.conflictType,
+            ring: this.conflict.ring,
+            attackers: this.conflict.attackers.slice(),
+            ringFate: this.conflict.ring.fate
+        };
+
+        this.game.raiseEvent(EventNames.OnConflictDeclaredBeforeProvinceReveal, params, event => {
+            if(this.conflict.attackers.some(a => a.location === Locations.PlayArea)) {
+                this.game.updateCurrentConflict(this.conflict);
+                this.conflict.declaredProvince = this.conflict.conflictProvince;
+                _.each(this.conflict.attackers, card => card.inConflict = true);
+                this.game.recordConflict(this.conflict);
+                const events = [];
+                if(this.conflict.ring.fate > 0 && this.conflict.attackingPlayer.checkRestrictions('takeFateFromRings', this.game.getFrameworkContext())) {
+                    this.game.addMessage('{0} takes {1} fate from {2}', this.conflict.attackingPlayer, this.conflict.ring.fate, this.conflict.ring);
+                    this.game.actions.takeFateFromRing({
+                        origin: this.conflict.ring,
+                        recipient: this.conflict.attackingPlayer,
+                        amount: this.conflict.ring.fate
+                    }).addEventsToArray(events, this.game.getFrameworkContext(this.conflict.attackingPlayer));
+                }
+                events.push(this.game.getEvent(EventNames.Unnamed, {}, () => {
+                    this.game.queueSimpleStep(() => this.promptForCovert());
+                    this.game.queueSimpleStep(() => this.resolveCovert());
+                }));
+                this.game.openThenEventWindow(events);
```

Comment:
I think we need to raise a Crashing Wave event here, but it seems weird to open two event windows like this in a row...